### PR TITLE
CPR-142 add retry dlq cronjob with support for scheduled downtime

### PR DESCRIPTION
**Pull request by markreesmoj**
This is the output with `retryDlqCronjob` and `scheduledDowntime` enabled

```# Source: hmpps-person-record/charts/generic-service/templates/retry-dlq-cronjob.yaml
apiVersion: batch/v1
kind: CronJob
metadata:
  name: hmpps-person-record-queue-housekeeping-cronjob
  labels:
    helm.sh/chart: generic-service-2.10.1
    app: hmpps-person-record
    release: hmpps-person-record
    app.kubernetes.io/version: "latest"
    app.kubernetes.io/managed-by: Helm
spec:
  schedule: "*/10 7-21 * * 1-5"
  concurrencyPolicy: Forbid
  failedJobsHistoryLimit: 5
  startingDeadlineSeconds: 600
  successfulJobsHistoryLimit: 5
  jobTemplate:
    spec:
      template:
        spec:
          serviceAccountName: person-record-service
          containers:
            - name: housekeeping
              image: ghcr.io/ministryofjustice/hmpps-devops-tools
              args:
                - /bin/sh
                - -c
                - curl -XPUT --connect-timeout 5 --max-time 10 --retry 5 --retry-delay 0 --retry-max-time 40 http://hmpps-person-record/queue-admin/retry-all-dlqs
          restartPolicy: Never```

### CODE REVIEW

File: charts/generic-service/templates/retry-dlq-cronjob.yaml [hunk 1]
```diff
@@ -0,0 +1,33 @@
+{{- if .Values.retryDlqCronjob.enabled -}}
+apiVersion: batch/v1
+kind: CronJob
+metadata:
+  name: {{ include "generic-service.fullname" . | trunc 26 }}-queue-housekeeping-cronjob
+  labels:
+    {{- include "generic-service.labels" . | nindent 4 }}
+spec:
+  {{- if .Values.scheduledDowntime.enabled }}
+  schedule: "{{ .Values.scheduledDowntime.retryDlqSchedule }}"
```
Comment (markreesmoj) [inline charts/generic-service/templates/retry-dlq-cronjob.yaml:10]: double quotes required here

File: charts/generic-service/values.yaml [hunk 2]
```diff
@@ -219,3 +219,9 @@ scheduledDowntime:
   # Stop at 10pm UTC Monday-Friday
   shutdown: '0 22 * * 1-5'
   serviceAccountName: scheduled-downtime-serviceaccount
+  # every 10 minutes between 7am and 10pm UTC Monday-Friday
```
Comment (markreesmoj) [inline charts/generic-service/values.yaml:222]: this means there will be no retries for an hour after restart - not sure how much of a big deal this is but thought it better to do this than risk a failing job by starting it at 6.

File: charts/generic-service/templates/retry-dlq-cronjob.yaml [hunk 3]
```diff
@@ -0,0 +1,33 @@
+{{- if .Values.retryDlqCronjob.enabled -}}
+apiVersion: batch/v1
+kind: CronJob
+metadata:
+  name: {{ include "generic-service.fullname" . | trunc 26 }}-queue-housekeeping-cronjob
```
[the remark this answers is not in the document]
Comment (petergphillips) [reply charts/generic-service/templates/retry-dlq-cronjob.yaml:5]: not a great fan of calling it a queue housekeeping cronjob, housekeeping doesn't really mean much to me 😁 
```suggestion
  name: {{ include "generic-service.fullname" . | trunc 26 }}-retry-dlqs
```

File: charts/generic-service/templates/retry-dlq-cronjob.yaml [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  concurrencyPolicy: Forbid
+  failedJobsHistoryLimit: 5
+  startingDeadlineSeconds: 600
+  successfulJobsHistoryLimit: 5
+  jobTemplate:
+    spec:
+      template:
+        spec:
+          {{- if .Values.serviceAccountName }}
+          serviceAccountName: {{ .Values.serviceAccountName }}
```
Comment (petergphillips) [inline charts/generic-service/templates/retry-dlq-cronjob.yaml:23]: odd to be using the service account to make a `curl` request to a running container?

File: charts/generic-service/templates/retry-dlq-cronjob.yaml [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+          {{- end }}
+          containers:
+            - name: housekeeping
```
Comment (petergphillips) [inline charts/generic-service/templates/retry-dlq-cronjob.yaml:26]:
```suggestion
            - name: retry-all-dlqs
```

File: charts/generic-service/templates/retry-dlq-cronjob.yaml [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+              image: ghcr.io/ministryofjustice/hmpps-devops-tools
```
Comment (mattops) [inline charts/generic-service/templates/retry-dlq-cronjob.yaml:27]: This should be a configurable value.

Comment (mattops) [reply charts/generic-service/templates/retry-dlq-cronjob.yaml:27]: e.g. ```
retryDlqCronjob:
  image: ghcr.io/ministryofjustice/hmpps-devops-tools
  tag: latest

File: charts/generic-service/templates/retry-dlq-cronjob.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (markreesmoj) [reply charts/generic-service/templates/retry-dlq-cronjob.yaml:5]: https://github.com/search?q=org%3Aministryofjustice+-retry-dlqs&type=code
vs
https://github.com/search?q=org%3Aministryofjustice%20-queue-housekeeping-cronjob&type=code
Github is 3-2 in favour of -queue-housekeeping-cronjob but we have the power to change that :-)

File: charts/generic-service/Chart.yaml [hunk 7]
```diff
@@ -7,4 +7,4 @@ type: application
 # This is the chart version. This version number should be incremented each time you make changes
 # to the chart and its templates, including the app version.
 # Versions are expected to follow Semantic Versioning (https://semver.org/)
-version: 2.9.2
+version: 2.10.1
```
[the remark this answers is not in the document]
Comment (markreesmoj) [reply charts/generic-service/Chart.yaml:10]: As per Matt's comment in Slack I have made this a patch, so 2.9.3

File: charts/generic-service/templates/retry-dlq-cronjob.yaml (hunk 4 above)
Comment (mattops) [reply charts/generic-service/templates/retry-dlq-cronjob.yaml:23]: if you use ` {{ .Values.serviceAccountName }}` then you may give this pod a lot more privileges than it requires.  You could create another value `{{ .Values. retryDlqCronjob.serviceAccountName }}`

File: charts/generic-service/templates/retry-dlq-cronjob.yaml (hunk 6 above)
Comment (markreesmoj) [reply charts/generic-service/templates/retry-dlq-cronjob.yaml]: 👍

**Review by petergphillips — APPROVED** [review]
LGTM

**Review by mattops — APPROVED** [review]

**Pull request merged**